### ISSUE TITLE
Give `log()` optional `base` argument

### ISSUE THREAD

**Issue by Arecher** [godotengine/godot-proposals#6652]
### Describe the project you are working on

Simply needed to do some simple math. In this case I wanted to solve `8 = 2^x` kind of scenario. 

### Describe the problem or limitation you are having in your project

I remembered needing to use `log()` to calculate the exponent, but when filling in `log(8)`, that obviously didn't give the desired answer. Obviously I needed to include the base for the log, but I couldn't find a way to do this at all.

This seemed like such a standard (and frankly mandatory) use of the `log()` GDScript function, I couldn't understand that it wasn't part of the standard function. I kept scouring the documentation for an alternative function that did allow me to specify a base, but couldn't find it at all. Turns out it isn't a thing at all.

Now to most seasoned programmers, it makes sense to immediately know how to work around this. (`log(8) / log(2)` does the trick), but to someone that hasn't used this in years, knowing that I need to use a base, but not getting any way to specify it, felt extremely strange and very confusing. This felt like a common use-case, and one that should be supported by Godot, especially for new users (or ones that have not needed to do such math in years).

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

The way I see it, there are two possible solutions:

1. Add an optional `base` argument to the log() function. This will allow anyone to easily understand how to use log(), even if they need to use a specific base, without needing to know how to calculate this manually. Making it an optional argument would avoid breaking compatibility.

2. The documentation for `log()` should mention the common use case of specifying a base, and how to calculate this, in case a user doesn't know how to do this. I presume this is often the case for users that need to look up `log()` in the documentation anyway.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

`log(float s)` would become `log(float s, float base)`

### If this enhancement will not be used often, can it be worked around with a few lines of script?

As said above, it can totally be worked around easily. But only if you know the math of how to do this. And if you do not know how to do it, it can take a while of googling before you know how to do it. Just like all of the other standard GDScript functions, they can easily be worked around with some simple math. But the whole point of them, is to make scripting user friendly, and easy to use. If you need to use a base, you should be allowed to do so, in the way that any user would expect.

This has been discussed [before](https://github.com/godotengine/godot/issues/9317), before it was closed due to moving from Godot 2 to Godot 3, but as far as I can see, hasn't been requested since. 

In the above linked old issue, someone mentions adding such a feature would clutter the API. I cannot see how this would clutter significantly, nor how that would be a good argument to not improve the user experience, over what seems like a relatively simple expansion to a core math function.

### Is there a reason why this should be core and not an add-on in the asset library?

It's a core feature of a GDScript function.

**Comment by Calinou** [comment]
- See https://github.com/godotengine/godot/pull/75300.

**Comment by Arecher** [comment]
Seems I'm two weeks late to the party! The documentation update would mitigate this issue greatly. As long as it means it's also updated for the 3.x documentation as well.

**Comment by Calinou** [comment]
I think it makes sense to add an optional `base` argument to log. If we can do it without breaking compatibility, I don't see why not.

**Comment by AThousandShips** [comment]
To avoid risking compat breaking we can just add `logb` with a specific base, as detection of the default value could be tricky (while some confusion from assumptions that `logb` should be of base 2 could happen)

**Comment by JB-ER** [comment]
I would really like to have this feature, because writing an util function for every project that uses some kind of math is a bit annoying and it just doesn't feel right to have a custom function for such a simple thing.

**Comment by EdwardChanCH** [comment]
I investigated/ partially implemented this yesterday and found some problems:

> I still think it makes sense to add an optional `base` argument to log – if we can do it without breaking compatibility, which is likely possible.

This would require adding a lot of boilerplate code and refactoring 8+ files. Unlike `VariantBuiltInMethod`, `VariantUtilityFunction` does not have default arguments as a parameter in all related call/ helper functions. `call_utility_function()` is also not a template, so I cannot make a new const array to tag on the default arguments (unlike `call_with_variant_args_static_ret_dv()`). 

<img width="591" alt="Image" src="https://github.com/user-attachments/assets/5dcfeb0a-154b-4040-b919-05202e514bfe" />

Method overloading is not an option (until #1571 is implemented), since `has_utility_function()` looks up a unique method name. 

> To avoid risking compat breaking we can just add logb with a specific base, as detection of the default value could be tricky (while some confusion from assumptions that logb should be of base 2 could happen)

This would break compatibility, since global functions hide local functions. 

<img width="540" alt="Image" src="https://github.com/user-attachments/assets/e2d831dc-e3e7-4268-bd0f-1cbd2ecd471a" />

But this would be the easiest option if compatibility were not an issue. 

Though, maybe use a different name:
- logb() means `log r (|x|)` in [C++](https://en.cppreference.com/w/cpp/numeric/math/logb) (usually r = 2).
- logn() sounds more intuitive, though it is used differently in [Informix Servers](https://www.ibm.com/docs/en/informix-servers/12.10.0?topic=routines-logn-function), and tends to get auto-corrected to `long()` or `login()`.
- logz() is an AI company.
- log2(), log10(), log1p() are commonly used in both [Python](https://docs.python.org/3/library/math.html#math.log10) and [C++](https://en.cppreference.com/w/cpp/numeric/math/log10).
- log_base() is the clearest, if not a bit long to type.

**Comment by EdwardChanCH** [comment]
Relevant links: 

- Someone implemented default arguments in utility functions 2 months ago: godotengine/godot#95017 
- The above solves a similar issue of adding default arguments to `is_equal_approx()` and `is_zero_approx()`: #4425 
- Changing the number of arguments in utility functions breaks binary compatibility with GDExtension (and requires more refactoring): https://github.com/godotengine/godot/pull/104408#issuecomment-2777719833

In summary, adding an optional argument to `log()` is preferred over adding a dedicated `log_base()`, but that requires a lot more refactoring. Adding `log2()`, `log10()`, and `log1p()` (as in Python/ C++) is simpler, and they are more efficient than using log(x, base) in most situations.
